packages/dds/tree/src/test/rebaserAxiomaticTests.ts

```diff
@@ -380,6 +654,196 @@ function verifyComposeAssociativity<TChangeset>(
+	};
+
+	// (A ○ B) ↷ C
+	const actualChange = rebaseTagged(
+		fieldRebaser.rebase,
+		makeAnonChange(fieldRebaser.compose(editA, editB)),
+		[editC],
+	);
+
+	const editRebasedOverC = rebaseTagged(fieldRebaser.rebase, editA, [editC]);
+	const invertedEdit = tagRollbackInverse(fieldRebaser.invert(editA), undefined, editA.revision);
+	// (A ↷ C) ○ (B ↷ (A⁻¹ ○ C ○ (A ↷ C)))
+	const expectedChange = makeAnonChange(
+		fieldRebaser.compose(
+			editRebasedOverC,
+			rebaseTagged(fieldRebaser.rebase, editB, [
+				makeAnonChange(
+					fieldRebaser.compose(
+						makeAnonChange(fieldRebaser.compose(invertedEdit, editC)),
+						editRebasedOverC,
+					),
+				),
+			]),
+		),
+	);
+
+	assertDeepEqual(actualChange, expectedChange);
+}
+
+function verifyRebaseOverUndoRedoPair<TChangeset>(
+	edit: TaggedChange<TChangeset>,
+	namedEditsToRebaseOver: NamedChangeset<TChangeset>[],
+	fieldRebaser: BoundFieldChangeRebaser<TChangeset>,
+) {
+	const assertDeepEqual = getDefaultedEqualityAssert(fieldRebaser);
+	assert(namedEditsToRebaseOver.length === 1, "expected just 1 edit");
+	const editsToRebaseOver = namedEditsToRebaseOver.map(({ changeset }) => changeset);
+	const editB = { change: editsToRebaseOver[0].change, revision: editsToRebaseOver[0].revision };
+
+	const inverseEditB = fieldRebaser.invert(editB);
+
+	// ((A ↷ B) ↷ B⁻¹) ↷ B
+	const actualChange = makeAnonChange(
+		fieldRebaser.rebase(
+			fieldRebaser.rebase(
+				fieldRebaser.rebase(edit.change, editB),
+				tagChange(inverseEditB, editB.revision),
+			),
+			editB,
+		),
+	);
+
+	// A ↷ B
+	const expectedChange = makeAnonChange(fieldRebaser.rebase(edit.change, editB));
+	assertDeepEqual(actualChange, expectedChange);
+}
+
+function verifyRebaseOverDoUndoPairIsNoOp<TChangeset>(
+	edit: TaggedChange<TChangeset>,
+	namedEditsToRebaseOver: NamedChangeset<TChangeset>[],
+	fieldRebaser: BoundFieldChangeRebaser<TChangeset>,
+) {
+	const assertDeepEqual = getDefaultedEqualityAssert(fieldRebaser);
+	assert(namedEditsToRebaseOver.length === 1, "expected just 1 edit");
+	const editsToRebaseOver = namedEditsToRebaseOver.map(({ changeset }) => changeset);
+
+	const editB = { change: editsToRebaseOver[0].change, revision: editsToRebaseOver[0].revision };
+
+	const invertedEditB = fieldRebaser.invert(editB);
+	// (A ↷ B) ↷ B⁻¹
+	const actualChange = makeAnonChange(
+		fieldRebaser.rebase(
+			fieldRebaser.rebase(edit.change, editB),
+			tagChange(invertedEditB, editB.revision),
+		),
+	);
+
+	const expectedChange = makeAnonChange(edit.change);
+	if (fieldRebaser.compareWithoutTombstones !== undefined) {
+		fieldRebaser.compareWithoutTombstones(actualChange.change, expectedChange.change);
+	} else {
+		assertDeepEqual(actualChange, expectedChange);
+	}
+}
+
+function verifyRebaseOverEmpty<TChangeset>(
+	edit: TaggedChange<TChangeset>,
+	fieldRebaser: BoundFieldChangeRebaser<TChangeset>,
+) {
+	const emptyChange = tagChange(
+		fieldRebaser.createEmpty(),
+		"test" as unknown as SessionSpaceCompressedId,
```

Comment:
This could be a global constant and shared with other tests.

Reply:
Using global constant shared with other tests